pkg/registryserver/registryserver.go

```diff
@@ -74,11 +86,17 @@ func (rs *RegistryServer) RegisterController(ctx context.Context, req *registry.
 	}
 	glog.V(3).Infof("Registering node: %s, endpoint: %s", req.NodeId, req.Endpoint)
 
-	rs.nodeClients[req.NodeId] = &NodeInfo{
+	node := &NodeInfo{
 		NodeID:   req.NodeId,
 		Endpoint: req.Endpoint,
 	}
 
+	rs.nodeClients[req.NodeId] = node
+
+	for l := range rs.listeners {
+		l.OnNodeAdded(ctx, node)
+	}
```

Comment:
As discussed on chat, this code deadlocks:
`RegisterController()->OnNodeAdded()->ConnectToNodeController()->GetNodeController()`

If `rs.listeners` doesn't need to be locked because it is only set once during initialization *and* the callback code itself is resilient against race conditions, then you can do:
```
rs.mutex.Lock()
defer func() {
    rs.mutex.Unlock()
    for l := range rs.listeners {
		l.OnNodeAdded(ctx, node)
    }
} ()
```

Reply:
I would prefer to lock and unlock only where needed, instead of locking on beginning of the function:

```
rs.mutex.Lock()
rs.nodeClients[req.NodeId] = node
rs.mutex.Unlock()
```

Reply:
I don't like that because it is a less effective protection against race conditions - see my other comment about `UnregisterController`.

But if you prefer this, then so be it.